formats/src/main/scala/org/scanamo/DynamoArray.scala

```diff
@@ -0,0 +1,220 @@
+    case PureN(xs)   => new AttributeValue().withNS(unsafeToList(xs, identity[String]))
+    case PureB(xs)   => new AttributeValue().withBS(unsafeToList(xs, identity[ByteBuffer]))
+    case Concat(xs, ys) =>
+      new AttributeValue().withL(unsafeMerge[AttributeValue](xs.toJavaCollection, ys.toJavaCollection))
+  }
+
+  def toJavaCollection: ju.List[AttributeValue] = self match {
+    case Empty      => new java.util.LinkedList[AttributeValue]()
+    case Strict(xs) => xs
+    case StrictS(xs) =>
+      xs.stream
+        .map[AttributeValue](jfun[String, AttributeValue](new AttributeValue().withS(_)))
+        .collect(Collectors.toList())
+    case StrictN(xs) =>
+      xs.stream
+        .map[AttributeValue](jfun[String, AttributeValue](new AttributeValue().withN(_)))
+        .collect(Collectors.toList())
+    case StrictB(xs) =>
+      xs.stream
+        .map[AttributeValue](jfun[ByteBuffer, AttributeValue](new AttributeValue().withB(_)))
+        .collect(Collectors.toList())
+    case Pure(xs)       => unsafeToList[DynamoValue, AttributeValue](xs, _.toAttributeValue)
+    case PureS(xs)      => unsafeToList[String, AttributeValue](xs, new AttributeValue().withS(_))
+    case PureN(xs)      => unsafeToList[String, AttributeValue](xs, new AttributeValue().withN(_))
+    case PureB(xs)      => unsafeToList[ByteBuffer, AttributeValue](xs, new AttributeValue().withB(_))
+    case Concat(xs, ys) => unsafeMerge(xs.toJavaCollection, ys.toJavaCollection)
+  }
+
+  def toDynamoValue: DynamoValue = DynamoValue.fromDynamoArray(self)
+
+  def isAttributeValueArray: Boolean = self match {
+    case _: Strict | _: Pure => true
+    case Concat(xs, ys)      => xs.isAttributeValueArray && ys.isAttributeValueArray
+    case _                   => false
+  }
+
+  def isStringArray: Boolean = self match {
+    case _: StrictS | _: PureS => true
+    case Concat(xs, ys)        => xs.isStringArray && ys.isStringArray
+    case _                     => false
+  }
+
+  def isNumArray: Boolean = self match {
+    case _: StrictN | _: PureN => true
+    case Concat(xs, ys)        => xs.isNumArray && ys.isNumArray
+    case _                     => false
+  }
+
+  def isByteBufferArray: Boolean = self match {
+    case _: StrictB | _: PureB => true
+    case Concat(xs, ys)        => xs.isByteBufferArray && ys.isByteBufferArray
+    case _                     => false
+  }
+
+  def asArray: Option[List[DynamoValue]] = self match {
+    case Empty => Some(List.empty)
+    case Strict(xs0) =>
+      Some(
+        xs0.stream.reduce[List[DynamoValue]](
+          Nil,
+          jfun2[List[DynamoValue], AttributeValue, List[DynamoValue]](
+            (xs, x) => xs :+ DynamoValue.fromAttributeValue(x)
+          ),
+          concatDynamoValues
+        )
+      )
+    case Pure(xs) => Some(xs)
+    case Concat(xs0, ys0) =>
+      for {
+        xs <- xs0.asArray
+        ys <- ys0.asArray
+      } yield xs ++ ys
```

Comment:
we use for comprehension a lot in place where a good old applicative would be enough. ie: `(xs0.asArray, ys0.asArray).mapN(_ ++ _)`
What do you think?

Reply:
Great idea!